### DISCUSSION TITLE
Can we customise the "output" field's formatting?

### DISCUSSION THREAD

**Discussion by Protofall** [gradescope/autograder_samples#62] [Q&A] [answered]
Hello, I'm sorry if this question has an obvious answer, but I was not able to find an answer in the documentation or via my own debugging over the past few days.

I've based my system on [this example](https://github.com/gradescope/autograder_samples/tree/master/diff_general), the key part is `test_generator.py` and the `assertIn` check. My expected output is:
```
1 2

```
So this has characters "1", "(space)", "2" and "newline-char" in that order, now if my program has the same output then the test passes and its all good, but my issue is with the fail case outputting. The `output` field in the JSON results file is set to things like

`"Test Failed: '1 2\\n' not found in 'invalid-output' : [msg]\n"`

Notice how the first newline char is escaped. This means on the gradescope website it actually outputs `\n` instead of doing a newline. I can see why this is the default behaviour, for most cases this is preferred, but I'd like to override this formatting and display the output in a more natural way, something like
```
Test failed:

Expected:
\```
1 2

\```

in user output:
\```
invalid-input
\```

[msg]
```

The problem is I don't know how and where to override the output string. I *think* this has nothing to do with the python unit tests/`assertIn` and instead is a gradescope-utils thing, but I don't know what exactly I'm looking for.

**Comment by liamsaliba** [comment, marked as answer]
The `assertIn` function (as well as other `assertX` functions) has a `msg` parameter where you can pass in a custom message.

To hide the default message generated by the `assertX` functions, you can call `self.longMessage = False` prior to calling the `assertX` function.  I've placed it inside of `setUp` in my own use.

**Reply by Protofall** [reply]
Ok, I knew about `msg` being appended to the default diff message, but I wasn't aware of `self.longMessage`. I'll try that out and will report back later.

**Reply by Protofall** [reply]
Yes, that's it! Thank you so much, I never would have found this flag on my own. For anyone whose curious, here's the "prettyAssert" function I made that gives the output I described above:

```
def prettyAssertIn(self, ex_out, out, bonus_msg):
    bar = "=========="

    msg = "\n\nExpected:\n"
    msg += bar + "\n"
    msg += ex_out
    msg += "\n" + bar
    msg += "\n\nIn user output:\n"
    msg += bar + "\n"
    msg += out
    msg += "\n" + bar + "\n"
    if bonus_msg != None:
        msg += "\n" + bonus_msg + "\n"

    longMessageCopy = self.longMessage
    self.longMessage = False
    self.assertIn(ex_out, out, msg=msg)
    self.longMessage = longMessageCopy
```
